packages/core/graph/src/AdjacencyList.js

```diff
@@ -0,0 +1,1195 @@
+    // the edge capacity in too large increments, which results in a lot of
+    // wasted memory.
+    let pct = capacity / PEAK_EDGE_CAPACITY;
+    let growFactor = interpolate(MAX_GROW_FACTOR, MIN_GROW_FACTOR, pct);
+    newCapacity = Math.round(capacity * growFactor);
+  } else if (currentLoadFactor < UNLOAD_FACTOR) {
+    // In some cases, it may be possible to shrink the edge capacity,
+    // but this is only likely to occur when a lot of edges have been removed.
+    newCapacity = Math.round(capacity * SHRINK_FACTOR);
+  }
+  assert(newCapacity <= MAX_EDGE_CAPACITY, 'Edge capacity overflow!');
+  return Math.max(MIN_EDGE_CAPACITY, newCapacity);
+}
+
+export default class AdjacencyList<TEdgeType: number = 1> {
+  /** An array of nodes, with each node occupying `NODE_SIZE` adjacent indices. */
+  #nodes: Uint32Array;
+  /** An array of edges, with each edge occupying `EDGE_SIZE` adjacent indices. */
+  #edges: Uint32Array;
+  /** A cache of connected nodes grouped by type. */
+  #typeMap: ?TypeMap<TEdgeType>;
+
+  constructor(
+    opts?: SerializedAdjacencyList<TEdgeType> | AdjacencyListOptions<TEdgeType>,
+  ) {
+    let nodes;
+    let edges;
+
+    if (opts?.nodes) {
+      // We were given a serialized adjacency list,
+      // so we just do a quick check of the data integrity
+      // and then initialize the `AdjacencyList`.
+      ({nodes, edges} = opts);
+      assert(
+        getNodesLength(nodes[CAPACITY]) === nodes.length,
+        'Node data appears corrupt.',
+      );
+
+      assert(
+        getEdgesLength(edges[CAPACITY]) === edges.length,
+        'Edge data appears corrupt.',
+      );
+    } else {
+      // We are creating a new `AdjacencyList` from scratch.
+      let {nodeCapacity = MIN_NODE_CAPACITY, edgeCapacity = MIN_EDGE_CAPACITY} =
+        opts ?? {};
+
+      assert(nodeCapacity <= MAX_NODE_CAPACITY, 'Node capacity is too large.');
+      assert(edgeCapacity <= MAX_EDGE_CAPACITY, 'Edge capacity is too large.');
+
+      // $FlowFixMe[incompatible-call]
+      nodes = new Uint32Array(
+        new SharedArrayBuffer(
+          getNodesLength(nodeCapacity) * Uint32Array.BYTES_PER_ELEMENT,
+        ),
+      );
+      nodes[CAPACITY] = nodeCapacity;
+
+      // $FlowFixMe[incompatible-call]
+      edges = new Uint32Array(
+        new SharedArrayBuffer(
+          getEdgesLength(edgeCapacity) * Uint32Array.BYTES_PER_ELEMENT,
+        ),
+      );
+      edges[CAPACITY] = edgeCapacity;
+    }
+
+    this.#nodes = nodes;
+    this.#edges = edges;
+  }
+
+  /**
+   * Create a new `AdjacencyList` from the given options.
+   */
+  static deserialize(
+    opts: SerializedAdjacencyList<TEdgeType>,
+  ): AdjacencyList<TEdgeType> {
+    return new AdjacencyList(opts);
+  }
+
+  /**
+   * Returns a serializable object of the nodes and edges in the graph.
+   */
+  serialize(): SerializedAdjacencyList<TEdgeType> {
+    return {
+      nodes: this.#nodes,
+      edges: this.#edges,
+    };
```

Comment:
The improvements to serialization and deserialization times come here; `nodes` and `edges` are both `SharedArrayBuffer` instances.